megatron/data/gpt_dataset.py

```diff
@@ -183,8 +188,44 @@ def __getitem__(self, idx):
                 self.doc_idx[doc_index_l],
                 length=offset_l + 1))
             sample = np.concatenate(sample_list)
+        
+        # Code from: https://github.com/EleutherAI/gpt-neox/blob/FIM-clean/megatron/data/gpt2_dataset.py#L109
+        # TODO(Hailey): can merge the code below this line with code above this line.
+        # TODO(Hailey), cont: above already iterates through loop, so just add the permuting in there?
+        sample = np.array(sample, dtype=np.int64)
+        # # print(sample, sample.shape)
+        # # do FIM here, if enabled
+        fim_rate = self.args.fim_rate
+
+        if fim_rate != 0:
+            assert (fim_rate <= 1 and fim_rate >= 0), "FIM rate must be a probability 0 <= rate <= 1"
+
+            eod = self.tokenizer.eod
+
+            segment_breaks = np.argwhere(sample == eod) # split sample by document
+
+            if segment_breaks.shape != (0, 1): # then there is an EOD token in this example
+                curr_start_position = 0
+                for loc in np.nditer(segment_breaks):
+                    # print(loc - curr_start_position, flush=True)
+                    # permute {prefix, suffix, middle} or {suffix, prefix, middle}
+                    # try:
+                    if loc - curr_start_position > 10: # sometimes examples start with EOD or are too short. so avoid this case
```

Comment:
Is there a specific reason for the minimum length 10?
What happens for examples smaller than that?

Reply:
In this case, I'm just not applying the FIM transformation to examples of less than 10 characters in length--the intuition was that either < 10-character documents would be 10 or fewer tokens and as such not be valuable to apply an infilling transformation to, or that in the process of applying the FIM transformation and adding 3 sentinel tokens, essentially all of the document (or all of the suffix portion, which is padded/truncated as needed [here](https://github.com/EleutherAI/gpt-neox/blob/f6eddc69b965da2d7cda8c8d3832f35af9ef4aee/megatron/data/gpt2_dataset.py#L378) ) would be lost thus making the document end up harming FIM performance.

The actual choice of 10 is arbitrary, and was not motivated by any #-of-tokens-to-characters factor w.r.t. the GPT-NeoX-20b tokenizer... but I hope this is helpful in whether you want to keep it in!

Reply:
(actually, this is a check for > 10 tokens in doc, my bad! but the reasoning was the same)